config/src/test/java/org/springframework/security/config/annotation/web/configurers/HeaderWriterLogoutHandlerTests.java

```diff
@@ -0,0 +1,137 @@
+@RunWith(SpringRunner.class)
+@SecurityTestExecutionListeners
+public class HeaderWriterLogoutHandlerTests {
+
+	private static final String CLEAR_SITE_DATA_HEADER = "Clear-Site-Data";
+
+	@Rule
+	public final SpringTestRule spring = new SpringTestRule();
+
+	@Autowired
+	MockMvc mvc;
+
+	@Test
+	@WithMockUser
+	public void logoutUsingClearSiteDataHandlerWithCsrfGet() throws Exception {
+		this.spring.register(HttpLogoutConfig.class).autowire();
+
+		this.mvc.perform(get("/logout").with(csrf()).secure(true))
+			.andExpect(result -> assertThat(result.getResponse().getHeader(CLEAR_SITE_DATA_HEADER))
+					.isNull());
+	}
+
+	@Test
+	@WithMockUser
+	public void logoutUsingClearSiteDataHandlerWithCsrfNotsecurePost() throws Exception {
+		this.spring.register(HttpLogoutConfig.class).autowire();
+
+		this.mvc.perform(post("/logout").with(csrf()))
+			.andExpect(result -> assertThat(result.getResponse().getHeader(CLEAR_SITE_DATA_HEADER))
+					.isNull());
+	}
+
+	@Test
+	@WithMockUser
+	public void logoutUsingClearSiteDataHandlerWithCsrfPost() throws Exception {
+		this.spring.register(HttpLogoutConfig.class).autowire();
+
+		this.mvc.perform(post("/logout").with(csrf()).secure(true))
+			.andExpect(result -> assertThat(result.getResponse().getHeader(CLEAR_SITE_DATA_HEADER))
+					.isEqualTo("\"cache\", \"cookies\", \"storage\", \"executionContexts\""));
+	}
+
+	@EnableWebSecurity
+	static class HttpLogoutConfig extends WebSecurityConfigurerAdapter {
+		@Override
+		protected void configure(HttpSecurity http) throws Exception {
+			http.logout()
+				.addLogoutHandler(new HeaderWriterLogoutHandler(
+						new ClearSiteDataHeaderWriter("cache", "cookies", "storage", "executionContexts")));
+		}
+	}
+
+	@Test
+	@WithMockUser
+	public void logoutUsingClearSiteDataHandlerWithoutCsrfGet() throws Exception {
+		this.spring.register(HttpLogoutConfigWithoutCSRF.class).autowire();
+
+		this.mvc.perform(get("/logout").secure(true))
+			.andExpect(result -> assertThat(result.getResponse().getHeader(CLEAR_SITE_DATA_HEADER))
+					.isEqualTo("\"storage\""));
+	}
+
+	@Test
+	@WithMockUser
+	public void logoutUsingClearSiteDataHandlerWithoutCsrfPost() throws Exception {
+		this.spring.register(HttpLogoutConfigWithoutCSRF.class).autowire();
+
+		this.mvc.perform(post("/logout").secure(true))
+			.andExpect(result -> assertThat(result.getResponse().getHeader(CLEAR_SITE_DATA_HEADER))
+					.isEqualTo("\"storage\""));
+	}
+
+	@Test
+	@WithMockUser
+	public void logoutUsingClearSiteDataHandlerWithoutCsrfNotSecure() throws Exception {
+		this.spring.register(HttpLogoutConfigWithoutCSRF.class).autowire();
+
+		this.mvc.perform(post("/logout"))
+			.andExpect(result -> assertThat(result.getResponse().getHeader(CLEAR_SITE_DATA_HEADER))
+					.isNull());
+	}
+
+	@EnableWebSecurity
+	static class HttpLogoutConfigWithoutCSRF extends WebSecurityConfigurerAdapter {
+		@Override
+		protected void configure(HttpSecurity http) throws Exception {
+			http.csrf().disable().logout()
```

Comment:
If you wouldn't mind, let's format this using [the formatting guidelines](https://spring.io/blog/2013/07/11/spring-security-java-config-preview-readability/), just for consistency:

```java
http
    .logout()
        .addLogoutHandler(...
```